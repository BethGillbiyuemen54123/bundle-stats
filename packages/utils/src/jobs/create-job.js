import {
  get, isEmpty, merge, set,
} from 'lodash';

import { SOURCE_PATH_WEBPACK_STATS, SOURCE_PATHS } from '../config';
import { createStats } from '../stats/create';
import { createStatsSummary } from '../stats/create-summary';
import {
  extractAssets,
  extractModules,
  extractModulesPackages,
  extractModulesPackagesDuplicate,
  extractMeta,
} from '../webpack';

/*
 * Create job from stats
 */
export const createJob = (source, baseline) => {
  const data = SOURCE_PATHS.reduce((agg, rawDataPath) => {
    const rawData = get(source, rawDataPath);

    if (!rawData) {
      return agg;
    }

    return merge(
      {},
      agg,
      {
        rawData: set({}, rawDataPath, rawData),
      },
    );
  }, {});

  const webpackData = get(data.rawData, SOURCE_PATH_WEBPACK_STATS);

  const { meta } = extractMeta(webpackData);
  const { assets } = extractAssets(webpackData);
  const { modules } = extractModules(webpackData);
  const { packages } = extractModulesPackages({ modules });

  const stats = createStats(baseline && baseline.rawData, data.rawData);
  const summary = createStatsSummary(baseline && baseline.stats, stats);

  const { warnings: duplicatePackagesWarnings } = extractModulesPackagesDuplicate({ packages });

  const warnings = {
    ...duplicatePackagesWarnings,
  };

  return {
    ...data,
    meta,
    stats,
    summary,
    ...isEmpty(warnings) ? {} : { warnings },
    metrics: {
      webpack: {
        assets,
        modules,
        packages,
      },
    },
  };
};
