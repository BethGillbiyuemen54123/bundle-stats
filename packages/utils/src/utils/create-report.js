import {
  flatMap, get, isEmpty, map, pick, uniq,
} from 'lodash';

import { METRIC_TYPE_FILE_SIZE } from '../config';
import * as webpack from '../webpack';
import { getMetricChanged, getMetricType, mergeRunsById } from '../metrics';
import { getDelta, formatDelta } from './delta';
import { formatPercentage } from './format';

export const addMetricsData = (entries, metricType, metricPrefix = '') => entries.map((entry) => {
  const { runs } = entry;
  const { biggerIsBetter, label, formatter } = getMetricType(
    metricPrefix ? [metricPrefix, entry.key].join('.') : entry.key,
    metricType,
  );

  return {
    ...entry,

    // Metric props
    biggerIsBetter,
    label,

    // Row props
    changed: getMetricChanged(runs),

    // Runs
    runs: runs.map((run, index) => {
      if (!run || typeof run.value === 'undefined' || run.value === null) {
        return null;
      }

      const { value, ...restRun } = run;
      const diff = (index < runs.length - 1) ? getDelta(runs[index + 1], run) : null;

      return {
        ...restRun,

        // run data
        value,
        displayValue: formatter(run.value),

        // diff data
        ...(diff !== null) ? {
          ...diff,
          displayDelta: formatDelta(diff.delta, formatter),
          displayDeltaPercentage: formatDelta(diff.deltaPercentage, formatPercentage),
        } : {},
      };
    }),
  };
});

const getWebpackSizesMetrics = (job) => {
  const metrics = get(job, 'metrics.webpack.sizes', {});

  return Object.entries(metrics).reduce((agg, [key, value]) => ({
    ...agg,
    [`webpack.sizes.${key}`]: value,
  }), {});
};

const getWebpackStatsMetrics = (job) => {
  const metrics = pick(get(job, 'metrics.webpack', {}), webpack.SUMMARY_METRIC_PATHS);

  return Object.entries(metrics).reduce((agg, [key, value]) => ({
    ...agg,
    [`webpack.${key}`]: value,
  }), {});
};

export const getModulesReport = (jobs) => map(
  uniq(flatMap(jobs, (job) => Object.keys(get(job, 'metrics.webpack.modules', {})))),

  (chunkId) => ({
    chunkId,
    chunkNames: uniq(flatMap(jobs, (job) => get(job, ['metrics', 'webpack', 'modules', chunkId, 'chunkNames']))),
    modules: addMetricsData(mergeRunsById(
      map(jobs, (job) => get(job, ['metrics', 'webpack', 'modules', chunkId, 'modules'])),
    ), METRIC_TYPE_FILE_SIZE),
  }),
);

export const createReport = (jobs) => {
  const warnings = get(jobs, '[0].warnings');

  return {
    runs: jobs.map(({ internalBuildNumber, meta }) => ({
      ...meta,
      internalBuildNumber,
    })),
    ...!isEmpty(warnings) ? { warnings } : {},
    stats: addMetricsData(mergeRunsById(map(jobs, getWebpackStatsMetrics))),
    sizes: addMetricsData(mergeRunsById(map(jobs, getWebpackSizesMetrics))),
    assets: addMetricsData(mergeRunsById(map(jobs, 'metrics.webpack.assets')), METRIC_TYPE_FILE_SIZE),
    modules: getModulesReport(jobs),
    packages: addMetricsData(mergeRunsById(map(jobs, 'metrics.webpack.packages')), METRIC_TYPE_FILE_SIZE),
  };
};
