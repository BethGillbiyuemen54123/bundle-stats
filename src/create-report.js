import path from 'path';
import { readFileSync } from 'fs-extra';
import { createJobs } from './create-jobs';

const TEMPLATE_FILEPATH = path.join(__dirname, './report.html');
const INITIAL_DATA_PATTERN = /"__INITIAL_DATA_PLACEHOLDER__"/;

export const createReport = (artifacts) => {
  const initialData = createJobs(artifacts);

  // @FIXME
  const template = readFileSync(TEMPLATE_FILEPATH, 'utf-8');
  return template.replace(INITIAL_DATA_PATTERN, JSON.stringify(initialData));
};
