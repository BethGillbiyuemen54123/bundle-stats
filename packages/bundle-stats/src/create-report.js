import { readFileSync } from 'fs-extra';
import { createReport } from '@bundle-stats/utils';

import {
  INITIAL_DATA_PATTERN,
  OUTPUT_FILENAME,
  OUTPUT_TYPE_HTML,
  OUTPUT_TYPE_JSON,
} from './constants';

const templateFilepath = require.resolve('@bundle-stats/html-templates');

export const createHTMLReport = (jobs) => {
  const template = readFileSync(templateFilepath, 'utf-8');
  return template.replace(INITIAL_DATA_PATTERN, `window.__INITIAL_DATA__ = ${JSON.stringify(jobs)}`);
};

export const createJSONReport = (jobs) => JSON.stringify(createReport(jobs), null, 2);

const REPORT_HANDLERS = {
  [OUTPUT_TYPE_HTML]: createHTMLReport,
  [OUTPUT_TYPE_JSON]: createJSONReport,
};

export const createReports = (jobs, options) => {
  const types = [
    ...options.html ? ['html'] : [],
    ...options.json ? ['json'] : [],
  ];

  return Promise.all(
    types.map((type) => ({
      output: REPORT_HANDLERS[type](jobs),
      filename: `${OUTPUT_FILENAME}.${type}`,
    })),
  );
};
