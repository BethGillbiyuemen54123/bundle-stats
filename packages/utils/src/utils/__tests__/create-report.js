import webpackStatsCurrentExtracted from '../../../__fixtures__/webpack-stats-1.extracted';
import webpackStatsBaselineExtracted from '../../../__fixtures__/webpack-stats-2.extracted';
import { createJobs } from '../../jobs';
import { createReport } from '../create-report';

describe('createReport', () => {
  test('single source', () => {
    const actual = createReport(createJobs([
      { webpack: webpackStatsCurrentExtracted },
    ]));

    expect(actual).toMatchSnapshot();
  });

  test('multiple sources', () => {
    const actual = createReport(createJobs([
      { webpack: webpackStatsCurrentExtracted },
      { webpack: webpackStatsBaselineExtracted },
    ]));

    expect(actual).toMatchSnapshot();
  });
});
