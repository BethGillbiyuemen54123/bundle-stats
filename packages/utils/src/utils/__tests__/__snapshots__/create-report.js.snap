// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createReport multiple sources 1`] = `
Object {
  "assets": Array [
    Object {
      "biggerIsBetter": false,
      "changed": true,
      "key": "main.css",
      "label": "main.css",
      "runs": Array [
        Object {
          "delta": -1000,
          "deltaPercentage": -9.09090909,
          "displayDelta": "-1000B",
          "displayDeltaPercentage": "-9.09%",
          "displayValue": "9.77KB",
          "isChunk": false,
          "isEntry": true,
          "isInitial": true,
          "name": "main.a1b2c3.css",
          "value": 10000,
        },
        Object {
          "displayValue": "10.74KB",
          "isChunk": false,
          "isEntry": true,
          "isInitial": true,
          "name": "main.b1c2d3.css",
          "value": 11000,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": true,
      "key": "main.js",
      "label": "main.js",
      "runs": Array [
        Object {
          "delta": 1000,
          "deltaPercentage": 2.04081633,
          "displayDelta": "+1000B",
          "displayDeltaPercentage": "+2.04%",
          "displayValue": "48.83KB",
          "isChunk": false,
          "isEntry": true,
          "isInitial": true,
          "name": "main.a2b3c4.js",
          "value": 50000,
        },
        Object {
          "displayValue": "47.85KB",
          "isChunk": false,
          "isEntry": true,
          "isInitial": true,
          "name": "main.b2c3d5.js",
          "value": 49000,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "logo.png",
      "label": "logo.png",
      "runs": Array [
        Object {
          "delta": 0,
          "deltaPercentage": 0,
          "displayDelta": "0B",
          "displayDeltaPercentage": "0%",
          "displayValue": "1000B",
          "isChunk": false,
          "isEntry": false,
          "isInitial": false,
          "name": "logo.a3b4c5.png",
          "value": 1000,
        },
        Object {
          "displayValue": "1000B",
          "isChunk": false,
          "isEntry": false,
          "isInitial": false,
          "name": "logo.a3b4c5.png",
          "value": 1000,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "index.html",
      "label": "index.html",
      "runs": Array [
        Object {
          "delta": 0,
          "deltaPercentage": 0,
          "displayDelta": "0B",
          "displayDeltaPercentage": "0%",
          "displayValue": "1.95KB",
          "isChunk": false,
          "isEntry": false,
          "isInitial": false,
          "name": "index.html",
          "value": 2000,
        },
        Object {
          "displayValue": "1.95KB",
          "isChunk": false,
          "isEntry": false,
          "isInitial": false,
          "name": "index.html",
          "value": 2000,
        },
      ],
      "type": undefined,
    },
  ],
  "modules": Array [
    Object {
      "chunkId": "1",
      "chunkNames": Array [
        "main",
      ],
      "modules": Array [
        Object {
          "biggerIsBetter": false,
          "changed": false,
          "key": "module-a",
          "label": "module-a",
          "runs": Array [
            Object {
              "delta": 0,
              "deltaPercentage": 0,
              "displayDelta": "0B",
              "displayDeltaPercentage": "0%",
              "displayValue": "1000B",
              "name": "module-a",
              "value": 1000,
            },
            Object {
              "displayValue": "1000B",
              "name": "module-a",
              "value": 1000,
            },
          ],
          "type": undefined,
        },
        Object {
          "biggerIsBetter": false,
          "changed": true,
          "key": "module-b",
          "label": "module-b",
          "runs": Array [
            Object {
              "delta": 500,
              "deltaPercentage": 33.33333333,
              "displayDelta": "+500B",
              "displayDeltaPercentage": "+33.33%",
              "displayValue": "1.95KB",
              "name": "module-b",
              "value": 2000,
            },
            Object {
              "displayValue": "1.46KB",
              "name": "module-b",
              "value": 1500,
            },
          ],
          "type": undefined,
        },
        Object {
          "biggerIsBetter": false,
          "changed": false,
          "key": "node_modules/package-a/index.js",
          "label": "node_modules/package-a/index.js",
          "runs": Array [
            Object {
              "delta": 0,
              "deltaPercentage": 0,
              "displayDelta": "0B",
              "displayDeltaPercentage": "0%",
              "displayValue": "1000B",
              "name": "node_modules/package-a/index.js",
              "value": 1000,
            },
            Object {
              "displayValue": "1000B",
              "name": "node_modules/package-a/index.js",
              "value": 1000,
            },
          ],
          "type": undefined,
        },
        Object {
          "biggerIsBetter": false,
          "changed": true,
          "key": "node_modules/package-a/node_modules/package-c/index.js",
          "label": "node_modules/package-a/node_modules/package-c/index.js",
          "runs": Array [
            Object {
              "delta": 1000,
              "deltaPercentage": 100,
              "displayDelta": "+1000B",
              "displayDeltaPercentage": "+100%",
              "displayValue": "1000B",
              "name": "node_modules/package-a/node_modules/package-c/index.js",
              "value": 1000,
            },
            null,
          ],
          "type": undefined,
        },
        Object {
          "biggerIsBetter": false,
          "changed": false,
          "key": "node_modules/package-b/index.js",
          "label": "node_modules/package-b/index.js",
          "runs": Array [
            Object {
              "delta": 0,
              "deltaPercentage": 0,
              "displayDelta": "0B",
              "displayDeltaPercentage": "0%",
              "displayValue": "1000B",
              "name": "node_modules/package-b/index.js",
              "value": 1000,
            },
            Object {
              "displayValue": "1000B",
              "name": "node_modules/package-b/index.js",
              "value": 1000,
            },
          ],
          "type": undefined,
        },
        Object {
          "biggerIsBetter": false,
          "changed": false,
          "key": "node_modules/package-c/index.js",
          "label": "node_modules/package-c/index.js",
          "runs": Array [
            Object {
              "delta": 0,
              "deltaPercentage": 0,
              "displayDelta": "0B",
              "displayDeltaPercentage": "0%",
              "displayValue": "1000B",
              "name": "node_modules/package-c/index.js",
              "value": 1000,
            },
            Object {
              "displayValue": "1000B",
              "name": "node_modules/package-c/index.js",
              "value": 1000,
            },
          ],
          "type": undefined,
        },
      ],
    },
  ],
  "packages": Array [
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "package-a",
      "label": "package-a",
      "runs": Array [
        Object {
          "delta": 0,
          "deltaPercentage": 0,
          "displayDelta": "0B",
          "displayDeltaPercentage": "0%",
          "displayValue": "1000B",
          "value": 1000,
        },
        Object {
          "displayValue": "1000B",
          "value": 1000,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": true,
      "key": "package-a:package-c",
      "label": "package-a:package-c",
      "runs": Array [
        Object {
          "delta": 1000,
          "deltaPercentage": 100,
          "displayDelta": "+1000B",
          "displayDeltaPercentage": "+100%",
          "displayValue": "1000B",
          "value": 1000,
        },
        null,
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "package-b",
      "label": "package-b",
      "runs": Array [
        Object {
          "delta": 0,
          "deltaPercentage": 0,
          "displayDelta": "0B",
          "displayDeltaPercentage": "0%",
          "displayValue": "1000B",
          "value": 1000,
        },
        Object {
          "displayValue": "1000B",
          "value": 1000,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "package-c",
      "label": "package-c",
      "runs": Array [
        Object {
          "delta": 0,
          "deltaPercentage": 0,
          "displayDelta": "0B",
          "displayDeltaPercentage": "0%",
          "displayValue": "1000B",
          "value": 1000,
        },
        Object {
          "displayValue": "1000B",
          "value": 1000,
        },
      ],
      "type": undefined,
    },
  ],
  "runs": Array [
    Object {
      "internalBuildNumber": 2,
      "webpack": Object {
        "builtAt": "2019-01-01T00:00:00.000Z",
        "hash": "abcd1234",
      },
    },
    Object {
      "internalBuildNumber": 1,
      "webpack": Object {
        "builtAt": "2019-01-01T00:00:00.000Z",
        "hash": "defg1234",
      },
    },
  ],
  "sizes": Array [
    Object {
      "biggerIsBetter": false,
      "changed": true,
      "key": "webpack.sizes.totalSizeByTypeCSS",
      "label": "CSS",
      "runs": Array [
        Object {
          "delta": -1000,
          "deltaPercentage": -9.09090909,
          "displayDelta": "-1000B",
          "displayDeltaPercentage": "-9.09%",
          "displayValue": "9.77KB",
          "value": 10000,
        },
        Object {
          "displayValue": "10.74KB",
          "value": 11000,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": true,
      "key": "webpack.sizes.totalSizeByTypeJS",
      "label": "JS",
      "runs": Array [
        Object {
          "delta": 1000,
          "deltaPercentage": 2.04081633,
          "displayDelta": "+1000B",
          "displayDeltaPercentage": "+2.04%",
          "displayValue": "48.83KB",
          "value": 50000,
        },
        Object {
          "displayValue": "47.85KB",
          "value": 49000,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "webpack.sizes.totalSizeByTypeIMG",
      "label": "IMG",
      "runs": Array [
        Object {
          "delta": 0,
          "deltaPercentage": 0,
          "displayDelta": "0B",
          "displayDeltaPercentage": "0%",
          "displayValue": "1000B",
          "value": 1000,
        },
        Object {
          "displayValue": "1000B",
          "value": 1000,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "webpack.sizes.totalSizeByTypeMEDIA",
      "label": "Media",
      "runs": Array [
        Object {
          "delta": 0,
          "deltaPercentage": 0,
          "displayDelta": "0B",
          "displayDeltaPercentage": "0%",
          "displayValue": "0B",
          "value": 0,
        },
        Object {
          "displayValue": "0B",
          "value": 0,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "webpack.sizes.totalSizeByTypeFONT",
      "label": "Fonts",
      "runs": Array [
        Object {
          "delta": 0,
          "deltaPercentage": 0,
          "displayDelta": "0B",
          "displayDeltaPercentage": "0%",
          "displayValue": "0B",
          "value": 0,
        },
        Object {
          "displayValue": "0B",
          "value": 0,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "webpack.sizes.totalSizeByTypeHTML",
      "label": "HTML",
      "runs": Array [
        Object {
          "delta": 0,
          "deltaPercentage": 0,
          "displayDelta": "0B",
          "displayDeltaPercentage": "0%",
          "displayValue": "1.95KB",
          "value": 2000,
        },
        Object {
          "displayValue": "1.95KB",
          "value": 2000,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "webpack.sizes.totalSizeByTypeOTHER",
      "label": "Other",
      "runs": Array [
        Object {
          "delta": 0,
          "deltaPercentage": 0,
          "displayDelta": "0B",
          "displayDeltaPercentage": "0%",
          "displayValue": "0B",
          "value": 0,
        },
        Object {
          "displayValue": "0B",
          "value": 0,
        },
      ],
      "type": undefined,
    },
  ],
  "stats": Array [
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "webpack.totalSizeByTypeALL",
      "label": "Total Bundle Size",
      "runs": Array [
        Object {
          "delta": 0,
          "deltaPercentage": 0,
          "displayDelta": "0B",
          "displayDeltaPercentage": "0%",
          "displayValue": "61.52KB",
          "value": 63000,
        },
        Object {
          "displayValue": "61.52KB",
          "value": 63000,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": true,
      "key": "webpack.totalInitialSizeJS",
      "label": "Initial JS Size",
      "runs": Array [
        Object {
          "delta": 1000,
          "deltaPercentage": 2.04081633,
          "displayDelta": "+1000B",
          "displayDeltaPercentage": "+2.04%",
          "displayValue": "48.83KB",
          "value": 50000,
        },
        Object {
          "displayValue": "47.85KB",
          "value": 49000,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": true,
      "key": "webpack.totalInitialSizeCSS",
      "label": "Initial CSS Size",
      "runs": Array [
        Object {
          "delta": -1000,
          "deltaPercentage": -9.09090909,
          "displayDelta": "-1000B",
          "displayDeltaPercentage": "-9.09%",
          "displayValue": "9.77KB",
          "value": 10000,
        },
        Object {
          "displayValue": "10.74KB",
          "value": 11000,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": true,
      "key": "webpack.cacheInvalidation",
      "label": "Cache Invalidation",
      "runs": Array [
        Object {
          "delta": 95.24,
          "deltaPercentage": 100,
          "displayDelta": "+95.24%",
          "displayDeltaPercentage": "+100%",
          "displayValue": "95.24%",
          "value": 95.24,
        },
        Object {
          "displayValue": "0%",
          "value": 0,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": true,
      "key": "webpack.moduleCount",
      "label": "Module Count",
      "runs": Array [
        Object {
          "delta": 1,
          "deltaPercentage": 20,
          "displayDelta": "+1",
          "displayDeltaPercentage": "+20%",
          "displayValue": "6",
          "value": 6,
        },
        Object {
          "displayValue": "5",
          "value": 5,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "webpack.chunkCount",
      "label": "Chunk Count",
      "runs": Array [
        Object {
          "delta": 0,
          "deltaPercentage": 0,
          "displayDelta": "0",
          "displayDeltaPercentage": "0%",
          "displayValue": "0",
          "value": 0,
        },
        Object {
          "displayValue": "0",
          "value": 0,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "webpack.assetCount",
      "label": "Asset Count",
      "runs": Array [
        Object {
          "delta": 0,
          "deltaPercentage": 0,
          "displayDelta": "0",
          "displayDeltaPercentage": "0%",
          "displayValue": "4",
          "value": 4,
        },
        Object {
          "displayValue": "4",
          "value": 4,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": true,
      "key": "webpack.packageCount",
      "label": "Package Count",
      "runs": Array [
        Object {
          "delta": 1,
          "deltaPercentage": 33.33333333,
          "displayDelta": "+1",
          "displayDeltaPercentage": "+33.33%",
          "displayValue": "4",
          "value": 4,
        },
        Object {
          "displayValue": "3",
          "value": 3,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": true,
      "key": "webpack.duplicatePackagesCount",
      "label": "Duplicate Packages Count",
      "runs": Array [
        Object {
          "delta": 1,
          "deltaPercentage": 100,
          "displayDelta": "+1",
          "displayDeltaPercentage": "+100%",
          "displayValue": "1",
          "value": 1,
        },
        Object {
          "displayValue": "0",
          "value": 0,
        },
      ],
      "type": undefined,
    },
  ],
  "warnings": Object {
    "webpack": Object {
      "duplicatePackages": Object {
        "package-c": Array [
          "package-a:package-c",
          "package-c",
        ],
      },
    },
  },
}
`;

exports[`createReport single source 1`] = `
Object {
  "assets": Array [
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "main.css",
      "label": "main.css",
      "runs": Array [
        Object {
          "displayValue": "9.77KB",
          "isChunk": false,
          "isEntry": true,
          "isInitial": true,
          "name": "main.a1b2c3.css",
          "value": 10000,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "main.js",
      "label": "main.js",
      "runs": Array [
        Object {
          "displayValue": "48.83KB",
          "isChunk": false,
          "isEntry": true,
          "isInitial": true,
          "name": "main.a2b3c4.js",
          "value": 50000,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "logo.png",
      "label": "logo.png",
      "runs": Array [
        Object {
          "displayValue": "1000B",
          "isChunk": false,
          "isEntry": false,
          "isInitial": false,
          "name": "logo.a3b4c5.png",
          "value": 1000,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "index.html",
      "label": "index.html",
      "runs": Array [
        Object {
          "displayValue": "1.95KB",
          "isChunk": false,
          "isEntry": false,
          "isInitial": false,
          "name": "index.html",
          "value": 2000,
        },
      ],
      "type": undefined,
    },
  ],
  "modules": Array [
    Object {
      "chunkId": "1",
      "chunkNames": Array [
        "main",
      ],
      "modules": Array [
        Object {
          "biggerIsBetter": false,
          "changed": false,
          "key": "module-a",
          "label": "module-a",
          "runs": Array [
            Object {
              "displayValue": "1000B",
              "name": "module-a",
              "value": 1000,
            },
          ],
          "type": undefined,
        },
        Object {
          "biggerIsBetter": false,
          "changed": false,
          "key": "module-b",
          "label": "module-b",
          "runs": Array [
            Object {
              "displayValue": "1.95KB",
              "name": "module-b",
              "value": 2000,
            },
          ],
          "type": undefined,
        },
        Object {
          "biggerIsBetter": false,
          "changed": false,
          "key": "node_modules/package-a/index.js",
          "label": "node_modules/package-a/index.js",
          "runs": Array [
            Object {
              "displayValue": "1000B",
              "name": "node_modules/package-a/index.js",
              "value": 1000,
            },
          ],
          "type": undefined,
        },
        Object {
          "biggerIsBetter": false,
          "changed": false,
          "key": "node_modules/package-a/node_modules/package-c/index.js",
          "label": "node_modules/package-a/node_modules/package-c/index.js",
          "runs": Array [
            Object {
              "displayValue": "1000B",
              "name": "node_modules/package-a/node_modules/package-c/index.js",
              "value": 1000,
            },
          ],
          "type": undefined,
        },
        Object {
          "biggerIsBetter": false,
          "changed": false,
          "key": "node_modules/package-b/index.js",
          "label": "node_modules/package-b/index.js",
          "runs": Array [
            Object {
              "displayValue": "1000B",
              "name": "node_modules/package-b/index.js",
              "value": 1000,
            },
          ],
          "type": undefined,
        },
        Object {
          "biggerIsBetter": false,
          "changed": false,
          "key": "node_modules/package-c/index.js",
          "label": "node_modules/package-c/index.js",
          "runs": Array [
            Object {
              "displayValue": "1000B",
              "name": "node_modules/package-c/index.js",
              "value": 1000,
            },
          ],
          "type": undefined,
        },
      ],
    },
  ],
  "packages": Array [
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "package-a",
      "label": "package-a",
      "runs": Array [
        Object {
          "displayValue": "1000B",
          "value": 1000,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "package-a:package-c",
      "label": "package-a:package-c",
      "runs": Array [
        Object {
          "displayValue": "1000B",
          "value": 1000,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "package-b",
      "label": "package-b",
      "runs": Array [
        Object {
          "displayValue": "1000B",
          "value": 1000,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "package-c",
      "label": "package-c",
      "runs": Array [
        Object {
          "displayValue": "1000B",
          "value": 1000,
        },
      ],
      "type": undefined,
    },
  ],
  "runs": Array [
    Object {
      "internalBuildNumber": 1,
      "webpack": Object {
        "builtAt": "2019-01-01T00:00:00.000Z",
        "hash": "abcd1234",
      },
    },
  ],
  "sizes": Array [
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "webpack.sizes.totalSizeByTypeCSS",
      "label": "CSS",
      "runs": Array [
        Object {
          "displayValue": "9.77KB",
          "value": 10000,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "webpack.sizes.totalSizeByTypeJS",
      "label": "JS",
      "runs": Array [
        Object {
          "displayValue": "48.83KB",
          "value": 50000,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "webpack.sizes.totalSizeByTypeIMG",
      "label": "IMG",
      "runs": Array [
        Object {
          "displayValue": "1000B",
          "value": 1000,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "webpack.sizes.totalSizeByTypeMEDIA",
      "label": "Media",
      "runs": Array [
        Object {
          "displayValue": "0B",
          "value": 0,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "webpack.sizes.totalSizeByTypeFONT",
      "label": "Fonts",
      "runs": Array [
        Object {
          "displayValue": "0B",
          "value": 0,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "webpack.sizes.totalSizeByTypeHTML",
      "label": "HTML",
      "runs": Array [
        Object {
          "displayValue": "1.95KB",
          "value": 2000,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "webpack.sizes.totalSizeByTypeOTHER",
      "label": "Other",
      "runs": Array [
        Object {
          "displayValue": "0B",
          "value": 0,
        },
      ],
      "type": undefined,
    },
  ],
  "stats": Array [
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "webpack.totalSizeByTypeALL",
      "label": "Total Bundle Size",
      "runs": Array [
        Object {
          "displayValue": "61.52KB",
          "value": 63000,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "webpack.totalInitialSizeJS",
      "label": "Initial JS Size",
      "runs": Array [
        Object {
          "displayValue": "48.83KB",
          "value": 50000,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "webpack.totalInitialSizeCSS",
      "label": "Initial CSS Size",
      "runs": Array [
        Object {
          "displayValue": "9.77KB",
          "value": 10000,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "webpack.cacheInvalidation",
      "label": "Cache Invalidation",
      "runs": Array [
        Object {
          "displayValue": "0%",
          "value": 0,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "webpack.moduleCount",
      "label": "Module Count",
      "runs": Array [
        Object {
          "displayValue": "6",
          "value": 6,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "webpack.chunkCount",
      "label": "Chunk Count",
      "runs": Array [
        Object {
          "displayValue": "0",
          "value": 0,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "webpack.assetCount",
      "label": "Asset Count",
      "runs": Array [
        Object {
          "displayValue": "4",
          "value": 4,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "webpack.packageCount",
      "label": "Package Count",
      "runs": Array [
        Object {
          "displayValue": "4",
          "value": 4,
        },
      ],
      "type": undefined,
    },
    Object {
      "biggerIsBetter": false,
      "changed": false,
      "key": "webpack.duplicatePackagesCount",
      "label": "Duplicate Packages Count",
      "runs": Array [
        Object {
          "displayValue": "1",
          "value": 1,
        },
      ],
      "type": undefined,
    },
  ],
  "warnings": Object {
    "webpack": Object {
      "duplicatePackages": Object {
        "package-c": Array [
          "package-a:package-c",
          "package-c",
        ],
      },
    },
  },
}
`;
